### PR TITLE
Move legacy meta boxes to toggle

### LINKED ISSUE #59246: Improve meta boxes in the Editor
Metaboxes in the Editor are unwieldy. While many plugins have been transitioning away from meta boxes, they remain an important tool for extensibility, and of course, the Custom Fields meta box remains in Core. 

Currently, when a meta box is enabled, it's displayed at the very bottom of the Editor. If you have a long post, you have to scroll all the way to the bottom to access the meta box. Here's an example from my personal site that utilizes Yoast SEO. 

yoast-metabox

Also when meta boxes are enabled, especially when starting a new post, they really clutter up the interface. 

image

It would be great to have another solution. Perhaps a new panel or modal that the metaboxes could be opened in. This would make it easier to access and manage meta box content, while also improving the Editor experience.

### LINKED ISSUE #61210: Typewriter Experience doesn't work when Custom Fields are shown
### Description

When the Custom Fields panel is shown, the Typewriter Experience / Typewriter Scrolling (#16460) doesn't work.

### Step-by-step reproduction instructions

1. In Editor Preferences on the Edit page, enable "Custom fields"
2. Save pref and reload the page
3. See that cursor is able to reach the bottom of the editor window, and can't be scrolled past

### Screenshots, screen recording, code snippet

_No response_

### Environment info

- WordPress 6.5.2
- Gutenberg version...unsure? Whatever is bundled with WP 6.5.2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

(would also fix #61210)

image

## Why?

Legacy meta boxes prevent the editor content from loading in an iframe. We cannot have two scrollable areas in the content area (and both the iframe and meta box area would be scrollable).

See #34105 for more background.

### Testing Instructions for Keyboard

Go to Options > Preferences > General > Advanced and toggle custom fields.

### CODE REVIEW

**Comment by ndiego** [comment]
I like this approach. It will actually make it much easier to edit metabox content.

**Comment by jasmussen** [comment]
Seems to balance the tradeoffs and allow for the best of both worlds. 

About this icon:

Screenshot 2024-02-21 at 16 00 26

I tried a new one:

Screenshot 2024-02-21 at 15 59 24

What do you think? If you're into it, here are the vectors, you can call it `metabox`:

```
<svg width="24" height="24" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M20 14v4c0 1.1-.9 2-2 2H6c-1.1 0-2-.9-2-2v-4h16Zm0-1.5V6c0-1.1-.9-2-2-2H6c-1.1 0-2 .9-2 2v6.5h1.5V6c0-.3.2-.5.5-.5h12c.3 0 .5.2.5.5v6.5H20ZM5.5 18c0 .3.2.5.5.5h12c.3 0 .5-.2.5-.5v-2.5h-13V18Z" /></svg>
```

**Comment by ellatrix** [comment]
@jasmussen The problem with that icon is that is symbolises the previous way we were showing meta boxes? Not that the icon I used it is great, I just didn't know what it use :)

**Comment by jasmussen** [comment]
How would you feel about this one instead?

Screenshot 2024-02-22 at 08 46 37

I personally prefer the one shared on this thread, but it's not a strong opinion.

**Comment by fabiankaegy** [comment]
I also really like this approach. But I just tested this with the first Plugin that came to mind still using the legacy metabox (Yoast) and it doesn't render correctly in there. They seem to render the metabox with a `is-hidden` class applied which only kicks in when the JS for it loads and hydrates their application. Which doesn't happen automatically anymore with this approach. 

I think this is a good test case we should be considering as the baseline test for this implementation.

**Comment by ellatrix** [comment]
What do you mean with Yoast not working? It seems to work for me:

Screenshot 2024-02-28 at 15 58 11

**Comment by fabiankaegy** [comment]
Okay I dont know what happened before. It is now working for me also. Thanks and sorry for the confusion

File: docs/reference-guides/data/data-core-edit-post.md [hunk 1]
```diff
@@ -328,6 +328,10 @@ _Returns_
 
 -   `boolean`: Whether the metaboxes are being saved.
 
+### showMetaBoxes
+
+Undocumented declaration.
```
Comment (ellatrix) [inline docs/reference-guides/data/data-core-edit-post.md:333]: Will make this private.

**Comment by jeffpaul** [comment]
Given the primary focus that the other buttons/icons in that area already have while creating content, might it be better to move this sort of toggle/interaction a bit farther away and perhaps into the `TOOLS` part of the `Options` dropdown (or perhaps even further into the `Preferences` menu)?  Being able to default to hiding those and not confusing editors by having that toggle so omnipresent among more important tasks would be my preference/recommendation here.

**Comment by ellatrix** [comment]
I'm open to that, but would love other people's input. Worth noting that this button will only appear if you have meta boxes. I do worry that it will become harder to access though.

File: packages/icons/src/library/metabox.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/**
+ * WordPress dependencies
+ */
+import { SVG, Path } from '@wordpress/primitives';
+
+const metabox = (
+	<SVG width="24" height="24" fill="none" xmlns="http://www.w3.org/2000/svg">
+		<Path
+			fill-rule="evenodd"
```
Comment (carolinan) [inline packages/icons/src/library/metabox.js:9]: When I apply the PR I see a JavaScript warning in the block editor:
`Warning: Invalid DOM property `fill-rule`. Did you mean `fillRule`?`

**Comment by mhkuu** [comment]
Hi 👋 I’m one of the lead devs at Yoast SEO. Could you please share a bit more of your plans here, including a planned release date? Could you please consider getting in touch with us to understand if there’s something we need to do on our side to stay backward and forward compatible? We really need to avoid situations like the 17.5 release that required a patch release for both [Gutenberg](https://github.com/WordPress/gutenberg/releases/tag/v17.5.1) and [Yoast SEO](https://github.com/Yoast/wordpress-seo/releases/tag/21.8.1) 😄

**Comment by fabiankaegy** [comment]
Hey @mhkuu 👋 

Thanks for chiming in here :) 

I cannot speak to any release timeline or more concise plans. The way these things tend to work with Gutenberg releases is that once a feature has recieved enough testing and a code review by a contributor it gets merged here and then released in the next Gutenberg release. (The release cadence is every 2 weeks with a RC every alternating week.)

As far as getting in touch about such updates I would just like to raise that I have created this issue in the Yoast repository in July of last year: https://github.com/Yoast/wordpress-seo/issues/20535. The addoption of the iFramed editor has been a slow rollout for the last 2 years now and to me this is the final step of getting us to that goal. 

Of course the UX is not ideal. I would much rather have plugins like Yoast addapt an actual native block editor integration that does not rely on the old legacy metabox implementation. 

If there are concerns / limitations with the native integration that are preventing you from addapting it / that are making it harder to addapt we are very open to recieving that feedback / making accommodations. But as it stands currently I am unaware of such callouts. 

As far as the best next steps here go. It would be great if you could test this PR with your plugin and provide detailed test feedback in case anything doesn't work as expected. As you can see in the comments above we have used the Yoast SEO plugin as a testcase ourselves and couldn't find any real issues in out tests.

**Comment by youknowriad** [comment]
> Of course the UX is not ideal. I would much rather have plugins like Yoast addapt an actual native block editor integration that does not rely on the old legacy metabox implementation.

I think we should try to emphasize this message more. The iframed editor is one reason but it's not the sole reason. As part of the admin redesign project, the site editor and post editor will become one at some point. I'm guessing the navigation won't require a page reload in these situations, so we need to ensure that metaboxes work in these contexts and it most likely means render them separately...
 
It's been more than 5 years since the release of WordPress 5.0. Plugins authors have had time to learn the APIs and we should make it clear that we favor these native integrations in the future. Maybe we should consider some outreach/P2 posts as well.

Now, for this particular PR, Its current target is WordPress 6.6 so ETA is July and probably sooner on the plugin to get more feedback and iron things out properly.

**Comment by arnaudbroes** [comment]
AIOSEO dev here. Just to confirm, the preferred approach is using a plugin sidebar?
https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-sidebar/

**Comment by fabiankaegy** [comment]
> AIOSEO dev here. Just to confirm, the preferred approach is using a plugin sidebar?
https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-sidebar/

@arnaudbroes well kind of. That is one of the slots avaliable. But it is't the only one. There are several slots to expose new panels / flows in the editor. All use [`registerPlugin`](https://github.com/WordPress/gutenberg/blob/trunk/packages/plugins/README.md#registerplugin) as their entry and from there any Slot can be used to surface features from the plugin.

**Comment by arnaudbroes** [comment]
> > AIOSEO dev here. Just to confirm, the preferred approach is using a plugin sidebar?
> > https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-sidebar/
> 
> @arnaudbroes well kind of. That is one of the slots avaliable. But it is't the only one. There are several slots to expose new panels / flows in the editor. All use [`registerPlugin`](https://github.com/WordPress/gutenberg/blob/trunk/packages/plugins/README.md#registerplugin) as their entry and from there any Slot can be used to surface features from the plugin.

Thanks for confirming. Correct, I just mentioned the sidebar since that's the main one we use but we use all other slots in our plugin as well.

Fortunately all of our features are already available through the sidebar (with modals) so it doesn't seem like we would need to do much to adjust the plugin to this change. I do hope that metaboxes will still remain accessible in one way or the other because I personally believe that our metabox offers the best UI/UX for our users.

**Comment by fabiankaegy** [comment]
> I do hope that metaboxes will still remain accessible in one way or the other because I personally believe that our metabox offers the best UI/UX for our users.

Do you have any feedback as to why you feel that way. Is it because there is not enough horizontal space available to build the UI's you are hoping in the available slots? 

What are the things that you feel make the non metabox approach less user friendly to you? :)

**Comment by arnaudbroes** [comment]
> Do you have any feedback as to why you feel that way. Is it because there is not enough horizontal space available to build the UI's you are hoping in the available slots?

Yep, that is the main reason. For instance, we have various features where we need to show a table with multiple columns of data to our users. Right now that obviously isn't possible in the sidebar considering we've have limited horizontal space there. To workaround this we open modals for these features if the user is using the sidebar. It works, but the metabox is obviously more user-friendly.

That being said I understand the underlying issue and I'm not really sure if there is any better way to approach this besides the solutions that have already been suggested. Unfortunately I'm not the best creative thinker.

I do think that enough time and thought needs to be invested into this since this change has the potential to upset a large amount of both users and plugin developers. I don't have numbers but I am certain that, even though we offer both a metabox and sidebar, an absolute majority of our users prefers using the metabox and I'm confident this is the same for many other plugins.

File: packages/icons/src/library/metabox.js (hunk 2 above)
Comment (ellatrix) [reply packages/icons/src/library/metabox.js:9]: Ah, thanks you :)

**Comment by ellatrix** [comment]
> For instance, we have various features where we need to show a table with multiple columns of data to our users.

Could you tell up more about what this is? Why not have a block and a visual representation?

**Comment by mhkuu** [comment]
There's a lot to unpack here. I feel that it might be worthwhile to consider a meeting with some stakeholders from the major plugins still relying (partly) on the metabox functionality. So I wholeheartedly agree with @arnaudbroes when they are saying:

> I do think that enough time and thought needs to be invested into this since this change has the potential to upset a large amount of both users and plugin developers. I don't have numbers but I am certain that, even though we offer both a metabox and sidebar, an absolute majority of our users prefers using the metabox and I'm confident this is the same for many other plugins.

I think as a relative newcomer in this environment a lot of things are over my head but I'm going to try and answer @fabiankaegy and @youknowriad as best as possible here. Also know that there is no ill intent.

## On the UX

> Of course the UX is not ideal. I would much rather have plugins like Yoast addapt an actual native block editor integration that does not rely on the old legacy metabox implementation.
> If there are concerns / limitations with the native integration that are preventing you from addapting it / that are making it harder to addapt we are very open to recieving that feedback / making accommodations. But as it stands currently I am unaware of such callouts.

Like AIOSEO, we have a sidebar integration (with modals), but arguably the metabox makes for a better editing experience. Moreover, very importantly for us in the plugin space, it enables discovery and thereby conversion. We can probably share some numbers on the use of the metabox vs. the sidebar in the block editor, if you're interested. 

> As far as the best next steps here go. It would be great if you could test this PR with your plugin and provide detailed test feedback in case anything doesn't work as expected. As you can see in the comments above we have used the Yoast SEO plugin as a testcase ourselves and couldn't find any real issues in out tests.

I've tried the implementation for a bit, and I will ask other developers and our QA engineers as well of course. In the short time I have spent testing, I have the feeling things generally work great (except for some CSS, also this warning `yoast-seo-metabox-css-css was added to the iframe incorrectly. Please use block.json or enqueue_block_assets to add styles to the iframe.​` which seems resolvable), so that's good news, props to everyone involved! 

However, the main pain point for me is that you can not see the metabox(es) and editor at the same time. With Yoast SEO, for example, that would mean that if you want to address something that's off in the content analysis, you would have to:

1. toggle the metabox
2. click the 'eye icon' in the metabox
3. close the metabox to return to the editor
4. see that some text is purple now
5. change that text
6. toggle the metabox
7. confirm that your issue is gone
8. close the metabox to return to the editor

See the video below:

https://github.com/WordPress/gutenberg/assets/7521233/a2df3aad-0b2f-4504-b464-9b549da69dfb

Wouldn't it be possible to provide a `SlotFill` below the editor so that developers could hook into that? 

This also aligns with these two comments (second from @arnaudbroes again):

> Do you have any feedback as to why you feel that way. Is it because there is not enough horizontal space available to build the UI's you are hoping in the available slots?

> Yep, that is the main reason. For instance, we have various features where we need to show a table with multiple columns of data to our users. Right now that obviously isn't possible in the sidebar considering we've have limited horizontal space there. To workaround this we open modals for these features if the user is using the sidebar. It works, but the metabox is obviously more user-friendly.

### Some side points 
- A very minor point is that I would consider another `aria-label` than `Show meta boxes`, not sure if a general user would know what those refer to.
- Non-developer focused, a minor point for us is that we will have to update a lot of test scenarios and video material to address this change. I guess that's just bad luck on our side and first-mover disadvantage and all but I do want you to understand that this change goes much broader than just development. 
- Note that it's not really an option to nuke our metabox altogether as we still need to support the classic editor. 

## On the release

> I cannot speak to any release timeline or more concise plans. The way these things tend to work with Gutenberg releases is that once a feature has recieved enough testing and a code review by a contributor it gets merged here and then released in the next Gutenberg release. (The release cadence is every 2 weeks with a RC every alternating week.)

Thanks, I know that's generally how this works around here, but given the importance of this change you might want to consider giving developers a little bit more of an advance warning. 

> Now, for this particular PR, Its current target is WordPress 6.6 so ETA is July and probably sooner on the plugin to get more feedback and iron things out properly.

The reason I need to know the Gutenberg release this lands on is that a lot of our WordPress.com customers get Gutenberg pre-installed on their environment with automatic updates turned on. That's why I referred to our shared patch releases (17.5.1 in your case). We definitely don't want to advice these customers to just deactivate Gutenberg 😅 

## On communication

> As far as getting in touch about such updates I would just like to raise that I have created this issue in the Yoast repository in July of last year: https://github.com/Yoast/wordpress-seo/issues/20535. The addoption of the iFramed editor has been a slow rollout for the last 2 years now and to me this is the final step of getting us to that goal.

We are aware of this issue in our repository, but you have to understand that we don't have all the context to understand the priority, timeline, and probably most importantly the "why". It might be super clear for you why you want to move to an iframed editor, but it might not be for us, so it's hard for us to plan accordingly, next to the myriad of other developments in Gutenberg and WordPress at large. 

> It's been more than 5 years since the release of WordPress 5.0. Plugins authors have had time to learn the APIs and we should make it clear that we favor these native integrations in the future. Maybe we should consider some outreach/P2 posts as well.

I don't know what P2 is, but you do need outreach indeed, coming back to the first paragraph of this comment. Also, I don't feel that this critique is warranted toward plugin developers, especially toward first-movers in this space.

It's WordCamp Asia now, maybe a good time for the lucky ones going to meet up and discuss? 😄

File: packages/edit-post/src/components/header/index.js [hunk 3]
```diff
@@ -189,6 +192,16 @@ function Header( { setEntitiesSavedStatesCallback, initialPost } ) {
 						setEntitiesSavedStatesCallback
 					}
 				/>
+				{ hasActiveMetaboxes && (
+					<Button
+						icon={ metabox }
+						onClick={ () => {
+							toggleMetaBoxes();
+						} }
+						label={ __( 'Show meta boxes' ) }
```
Comment (mhkuu) [inline packages/edit-post/src/components/header/index.js:201]: As pointed out, I would consider a different label as `meta boxes` might not be a term the general user is familiar with. Also, when you close the meta boxes you would want to have `Hide meta boxes` as the label, so in line with the `Settings` toggle for the sidebar I would consider going with just a noun here.

**Comment by youknowriad** [comment]
> Also, I don't feel that this critique is warranted toward plugin developers, especially toward first-movers in this space.

For clarify, this was not a critique at all, it was a suggestion for us as a community to start thinking about moving forward from Metaboxes in the editor now that there are alternative APIs and that enough time have been given to learn and embrace them.

**Comment by earnjam** [comment]
> > Also, I don't feel that this critique is warranted toward plugin developers, especially toward first-movers in this space.
> 
> For clarify, this was not a critique at all, it was a suggestion for us as a community to start thinking about moving forward from Metaboxes in the editor now that there are alternative APIs and that enough time have been given to learn and embrace them.

It was in a reply to a comment about Yoast specifically, so I think that is why it felt like a critique. We are fully aware of these APIs. Yoast SEO has utilized the plugin sidebar slotfill since it was first merged, and we have the same functionality of our metabox in our sidebar. Technical knowledge/understanding of block editor APIs is not the problem here. It's a UX problem. 

Despite multiple paths to surface the sidebar and full feature parity with the metabox, the overwhelming majority of our users still use (_and seem to prefer to use_) the metabox. We absolutely do not want to impeded core development, but it is very difficult to justify intentionally removing the metabox at the expense of the user experience of our customers. This exact [discussion came up on Twitter](https://x.com/fabiankaegy/status/1684884535232925696?s=20) when @fabiankaegy called it out last July, but nothing has really changed since then. 

And as @arnaudbroes pointed out, other developers of other prominent and widely used plugins see the same issues of limited space & user awareness. I personally prefer to use the more native-feel of the sidebars, but I am also not a good representation of an average user. I'm not sure how much of the lack of user adoption is a discoverability problem vs a usability problem. 

As far as this PR specifically: In addition to @mhkuu's specific feedback on having to toggle back and forth to interact with the metabox and the content, I'd add that we've seen other examples where hiding functionality behind generic icons [leads to user confusion and discoverability issues](https://github.com/WordPress/gutenberg/issues/55179). This feels similar. I can guarantee you that if this ships, we are going to field an increased support load of our users asking where their metabox has gone. I don't think we'd be alone there.

**Comment by annezazu** [comment]
To help with moving away from metaboxes and into the tooling we have today, I opened an issue to create a how to guide in the block editor specifically for this: https://github.com/WordPress/gutenberg/issues/59804

**Comment by bph** [comment]
I tested this and seems to be working fine as a POC. 

I see an advantage to make each plugin load their metaboxes into a separate modal. 

The ongoing scrolling if more than one plugin loads metaboxes for a post is quite cumbersome and a big downside of the current implementation. see video 

https://github.com/WordPress/gutenberg/assets/39980/4c381262-4fbc-4800-9a21-6c5070f4f59a

It would be an instant improvement if there were a separation of concern and each plugin use their own space. That way developers can have more controls over the layout and not step on each other's toes when design changes are implemented. 

The button could change into a drop-down box listing each plugin.

**Comment by youknowriad** [comment]
@bph thanks for testing. I don't think the metabox API allows us to discriminate per plugin. A plugin can register one metabox or multiple related meta boxes.

**Comment by bph** [comment]
Hmm @youknowriad It seems the Options > Preferences > Advanced screen is able to separate displaying of metaboxes by plugins. Maybe that's a start? 
Screenshot 2024-03-13 at 14 01 21
Screenshot 2024-03-13 at 14 01 03

**Comment by youknowriad** [comment]
@bph The problem is that is not a guarantee. It's worth considering but a plugin can register multiple meta boxes at the same time.

**Comment by youknowriad** [comment]
Screenshot 2024-03-13 at 2 10 30 PM
Screenshot 2024-03-13 at 2 10 40 PM

For instance ACF allows users to define field groups (to group fields together) but the expectation is that it's still the same form in the end.

**Comment by rilwis** [comment]
Hi, it's Anh from [Meta Box](https://metabox.io). Our plugin's goal is helping developers to create meta boxes. And this change might affect both us and our users.

One problem we see as @arnaudbroes said above about the UI. The current (below content) is a better place for meta boxes than sidebar slots, because of the horizontal space it can use. Our plugin allows users to put custom fields in columns, tabs and groups (including nested groups), so the UI can be complex for some use cases. The thing is while we can improve UI of each field, we leave the freedom to users to organize fields, so the UI is some kind of unpredictable.

In my quick test, the toggle seem to work fine for most fields. However, the [WYSIWYG](https://docs.metabox.io/fields/wysiwyg/) field doesn't allow users to enter content.

My concern (question) is: is there any technical issues that we need to worry about? I remember previously when Gutenberg started to support meta boxes, it didn't work with file uploads. FYI, we have more than 40 field types, from simple fields like text, textarea to complex fields like Google Maps, Open Street Maps, file upload, select2, date time picker, etc. If you can share more about the technical background here, so we can prepare for what's coming, that would be great.

**Review by Mamaduka — CHANGES_REQUESTED** [review]
While I agree with the proposed change, this is too much of a "sudden" swap for many developers, and we should keep compatibility in mind.

I understand the argument that metal boxes were always legacy APIs, but as a team focused on UX (I hope that's what we're doing), we can't disregard people using the canvas like this.

**Comment by youknowriad** [comment]
I don't disagree that we need to account for these plugins but as I explained before, we need to find a path forward for these plugins with the new APIs. 

Meta boxes already cause multiple regressions in the editor.

 - Frontend doesn't match backend.
 - Potentially two revisions per save (at least two requests).
 - They are not supported when editing "pages" in the site editor.

And may cause other regressions in the future.

 - May not be supported by the future post editor editor (With the new admin redesign refactor, the post editor and site editor will align). It's still unclear if it's possible to support yet, we'll have to do explorations here but even if we manage to make it work, it will be suboptimal because of the client side navigation. (This is not coming tomorrow but we have to prepare now)

So my thinking is what we can do to encourage developers to move away from meta-boxes now, which will allow us to make progress on the rest of the projects.

**Comment by Mamaduka** [comment]
Like I said, I don't disagree, but I think we need a better middle ground, but this isn't it (IMO). I don't have any ideas at the moment, but happy to help out.

P.S. I think we fixed the revision issue.

**Comment by arnaudbroes** [comment]
> Like I said, I don't disagree, but I think we need a better middle ground, but this isn't it (IMO). I don't have any ideas at the moment, but happy to help out.
> 
> P.S. I think we fixed the revision issue.

I completely agree with this, I fully understand that this change is needed to allow further progress to be made on other issues and features and I am not against making this change, but I don't think the current solution is ideal in terms of UX.

**Comment by youknowriad** [comment]
What should be the solution for you folks, what's preventing you from using the new APIs?

 - Plugin Sidebar
 - Document Panel
 - Blocks (connected blocks)

One other solution would be to auto-enable classic editor if we detect a meta-box but I don't really want to come to that.

**Comment by fabiankaegy** [comment]
@youknowriad I think the theme that I am reading through all the comments is that it's not a imitation of the available API's but rather a UX issue where manny plugins have interfaces that don't fit into a sidebar but also don't really fit the concept of blocks. 

A good example of this to me is how the current new experimental WooCommerce single product editing experience didn't opt to be fully block-based but instead uses a more fixed form approach to enter a lot of metadata. 

To me it sounds like the best way to get more plugins to migrate to using the block editor native approach would be to expose additional slots where UI can be rendered. Like introducing a plugin location at the bottom of the screen that does not use the old metabox API but actually uses native block editor apis. 

I know that @SaxonF has given this issue quite a bit of thought in the Admin Redesign for how the UI could handle forms in setting screens. Maybe some insights from those explorations could help us come up with a compromise that would make it easier for plugins to migrate over when their UI doesn't fit into a sidebar

**Comment by youknowriad** [comment]
That makes sense yeah!

> A good example of this to me is how the current new experimental WooCommerce single product editing experience didn't opt to be fully block-based but instead uses a more fixed form approach to enter a lot of metadata.

That's an interesting example and it shows that it's already possible to use block editor APIs and components to take over the screen and build such a custom editor.

> To me it sounds like the best way to get more plugins to migrate to using the block editor native approach would be to expose additional slots where UI can be rendered. Like introducing a plugin location at the bottom of the screen that does not use the old metabox API but actually uses native block editor apis.

I don't think new Slots are a solution because a meaningless slot (which is what meta boxes are) is just a way to hack the UI which will put us in the same situation again if we ever want to modify the UI again. We shouldn't repeat the mistakes IMO.

That said, I do think there's a path forward for things like data views, fields... which will address a lot of these use-cases but I don't really expect this to be ready soon while we may need a solution for meta boxes sooner.

**Comment by arnaudbroes** [comment]
> I think the theme that I am reading through all the comments is that it's not a imitation of the available API's but rather a UX issue where many plugins have interfaces that don't fit into a sidebar but also don't really fit the concept of blocks.

This basically sums it up. Many plugins like AIOSEO and Yoast have already been fully integrated with the sidebar and they use the aforementioned Block Editor components. It's just that we know that an absolute majority of our users prefer to use our metabox over the sidebar menu because the UI feels more intuitive. This is mostly due to the larger amount of horizontal space that is available (e.g. AIOSEO uses tables for a few features that simply do not fit in the sidebar, and we have to open a modal for those as a result).

> To me it sounds like the best way to get more plugins to migrate to using the block editor native approach would be to expose additional slots where UI can be rendered. Like introducing a plugin location at the bottom of the screen that does not use the old metabox API but actually uses native block editor apis.

If technically viable, that would indeed be ideal. We really don't mind having to adapt to new APIs, it's mostly a matter of having a similar amount of space available to us.

**Comment by ellatrix** [comment]
> To me it sounds like the best way to get more plugins to migrate to using the block editor native approach would be to expose additional slots where UI can be rendered. Like introducing a plugin location at the bottom of the screen that does not use the old metabox API but actually uses native block editor apis.

Even if we wanted to, it's not really possible to add a slot below the content without putting it _inside_ the iframe. If we add it outside the iframe, we end up with two scrollable areas because the iframe is scrollable and the meta boxes would need to be scrollable as well. We can't expand the iframe contents because that would break viewport units. For me two whole point of this PR is to have the iframed editor consistently regardless of whether a plugin uses meta boxes.

**Comment by ellatrix** [comment]
A slot within the iframe would mean it's all styled as content. So at that point you might as well create blocks?

**Comment by fabiankaegy** [comment]
@SaxonF actually just shared a new walkthrough of the current explorations for the admin redesign project in slack here: https://wordpress.slack.com/archives/C02S78ZAL/p1710741693260459?thread_ts=1710189675.902729&cid=C02S78ZAL

To me the idea of the content area which can be used to expose settings that are not directly tied to the frontend appearance of the site on the web makes a lot of sense. See the prototype: https://www.figma.com/proto/804HN2REV2iap2ytjRQ055/System?page-id=2244%3A16904&type=design&node-id=2737-40156&viewport=-2476%2C655%2C0.54&t=V97h1DsXrJBso2X6-1&scaling=min-zoom&starting-point-node-id=2737%3A40126&mode=design

And here one showcasing it populated with an example of a product CPT: https://www.figma.com/proto/804HN2REV2iap2ytjRQ055/System?page-id=2374%3A25967&type=design&node-id=2511-45456&viewport=748%2C1481%2C0.33&t=1lc4tsPo3i43E7Td-1&scaling=min-zoom&starting-point-node-id=2476%3A23475&show-proto-sidebar=1&mode=design

CleanShot 2024-03-18 at 19 57 33@2x


CleanShot 2024-03-18 at 20 01 15@2x

**Comment by gaambo** [comment]
> So at that point you might as well create blocks?

Just want to share my experience for creating custom blocks for "meta" data entry (not explicitly meta fields, but anything that is _about_ the content):
- It's technically absolutely doable by saving to meta in a block, that's not the problem.
- You have to put the block in the post types template and lock the block (move/remove)
- An admin (with canLockBlocks or access to the code editor) could still remove the block
- As a plugin developer providing such an interface, you can't make sure that a site builder / developer overwrites the editor template of a post type to only include certain blocks and template lock it. 
- The block still has a toolbar - even if moving/locking may be disabled, it still feels somewhat "wrong" to have that.
- The block will probably adhere to the theme's content/wide/full width. It makes it look like content. I think UIs that are _about_ the content, but not _content_ per se, should look different (and probably use @wordpress/components instead of the themes button styling etc.).

I've used it for a "testimonial" CPT, which only has a quote block + a custom block and a fully locked template - it works for that use case. But for a general plugin (like SEO) which will need to work for all post types, that may be suboptimal.

I understand the desire to provide more complex UIs to users as a plugin developer. The sidebar and its slots are great, but they mostly work for "vertical" forms with one input per row. Anything that's a table, needs a preview (like social sharing) etc. doesn't have enough space there. 

From an extenders' standpoint, I also think a SlotFill under the editor would be great. Although I see there are technical aspects that make that problematic. Some plugins building custom block editors also try to work with tabs at the top (like the new [WooCommerce](https://woo.com/document/new-product-editor-beta/#section-2) product editing experience, [givewp form builder](https://github.com/WordPress/gutenberg/assets/5585580/59033d8b-6c9f-4a7b-ab87-0ccc1a3c2818), LearnDash quiz builder). I don't think the UI is perfect, and I've got no idea if it's accessible. But from a user perspective it makes sense to have "content", "settings", at the top. 

Conclusion: I understand the direction this PR wants to take, and I support it - but I think this implementation may be a little too "harsh".

**Comment by zahardev** [comment]
Hi @ellatrix,

I'm a Seriously Simple Podcasting plugin developer.

Could you advise, will it work if we move the meta fields to the sidebar without implementing the toggle feature? Like this:

2024-04-15-bd192

**Comment by ndiego** [comment]
@ellatrix do you have any updates on the progress of this PR?

**Comment by ellatrix** [comment]
Definitely too late for this release now, let's try the next. I think I'll change the PR to a collapsible details panel in the footer area instead, to make the change less big.

**Comment by ndiego** [comment]
> Definitely too late for this release now, let's try the next. I think I'll change the PR to a collapsible details panel in the footer area instead, to make the change less big.

Sounds good, thanks for the update!

**Comment by bph** [comment]
This could use some testing 
[Collapse classic meta boxes under details element](https://github.com/WordPress/gutenberg/pull/64247#top)

**Comment by ellatrix** [comment]
Yes closing this in favour of #64247 :)

**Comment by ellatrix** [comment]
> The block will probably adhere to the theme's content/wide/full width. It makes it look like content. I think UIs that are about the content, but not content per se, should look different (and probably use @wordpress/components instead of the themes button styling etc.).


@gaambo You should not use blocks unless it's actually a preview of content as it would appear on the front-end. Blocks can use inspector controls and modals of course, which would be styled as UI.

**Comment by gaambo** [comment]
> > The block will probably adhere to the theme's content/wide/full width. It makes it look like content. I think UIs that are about the content, but not content per se, should look different (and probably use @wordpress/components instead of the themes button styling etc.).
> 
> @gaambo You should not use blocks unless it's actually a preview of content as it would appear on the front-end. Blocks can use inspector controls and modals of course, which would be styled as UI.

I know, but the [docs](https://developer.wordpress.org/block-editor/how-to-guides/metabox/) and other articles often mention to use blocks with a meta-field storage instead of meta boxes - it was mentioned in this discussion here as well. And I just wanted to chime into this discussion, that blocks - even if they store the value in the postmeta - are often not a good replacement for _meta_-boxes.

**Pull request closed without merging**